// partUserProfile/schema.js
exports.signUpPutHeader = {
    "type" : "object",
    "properties" : {
        "org" :{
            "type" : "string"
        }
    }
};

exports.signUpPutData = {
    "type": "object",
    "properties": {
        "data": {
            "type": "object",
            "properties": {
                "username": {
                    "type": "string"
                },
                "password": {
                    "type": "string"
                },
                "nationalCode": {
                    "type": "string"
                },
                "skill": {
                    "type": "string"
                },
                "job": {
                    "type": "string"
                },
                "managerName": {
                    "type": "string"
                },
                "firstName": {
                    "type": "string"
                },
                "lastName": {
                    "type": "string"
                },
                "gender": {
                    "type": "string"
                },
                "education": {
                    "type": "string"
                }
            },
            "additionalProperties": true,
            "required": [
                "username",
                "password",
                "nationalCode",
                "skill",
                "job",
                "managerName",
                "firstName",
                "lastName",
                "gender",
                "education"
            ]
        },
        "parent": {
            "type": "string"
        }
    },
    "additionalProperties": true,
    "required": [
        "data",
        "parent"
    ]
};

exports.loginHeader = {}

exports.loginData = {
    "type": "object",
    "properties": {
        "username": {
            "type": "string"
        },
        "password": {
            "type": "string"
        }
    },
    "additionalProperties": false,
    "required": [
        "username",
        "password"
    ]
}

exports.logoutHeader = {}

exports.logoutData = {}

exports.changePasswordHeader = {

}

exports.changePasswordData = {
    "type": "object",
    "properties": {
        "type": {
            "type": "string"
        },
        "username": {
            "type": "string"
        },
        "newPassword": {
            "type": "string"
        },
        "nationalCode": {
            "type": "string"
        },
        "email": {
            "type": "null"
        }
    },
    "additionalProperties": true,
    "required": [
        "newPassword",
        "nationalCode",
        "email"
    ]
}

exports.getProfileHeader = {}

exports.getProfileData = {}

exports.editProfileHeader = {}

exports.editProfileData = {
    "type": "object",
    "properties": {
        "uniqueKey": {
            "type": "null"
        },
        "insertDate": {
            "type": "string"
        },
        "registerType": {
            "type": "null"
        },
        "status": {
            "type": "null"
        },
        "type": {
            "type": "string"
        },
        "firstName": {
            "type": "string"
        },
        "lastName": {
            "type": "string"
        },
        "gender": {
            "type": "integer"
        },
        "maritalStatus": {
            "type": "null"
        },
        "fatherName": {
            "type": "null"
        },
        "identityNumber": {
            "type": "null"
        },
        "issuedFrom": {
            "type": "null"
        },
        "identitySerialNumber": {
            "type": "null"
        },
        "idNumber": {
            "type": "string"
        },
        "birthday": {
            "type": "null"
        },
        "education": {
            "type": "string"
        },
        "email": {
            "type": "null"
        },
        "cellphoneNumbers": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "addresses": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "tradingCode": {
            "type": "null"
        },
        "knowledgeLevel": {
            "type": "null"
        },
        "hasTrading": {
            "type": "null"
        },
        "brokerageCompanyName": {
            "type": "null"
        },
        "jobs": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "assets": {
            "type": "null"
        },
        "accounts": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "profilePictures": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "identityCards": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "idCards": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "accountStatements": {
            "type": "array",
            "items": {},
            "additionalItems": true
        },
        "signatureSamples": {
            "type": "array",
            "items": {},
            "additionalItems": true
        }
    },
    "additionalProperties": true,
    "required": [

    ]
}
