// partUserProfile/tests/userProfile/schema.js
exports.signUp = {
    type: 'object',
    properties: {
        data: {
            inArray: ["ثبت نام موفقیت امیز"]
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"},
                requestId:{type: "string"}
            }
        }
    }
}
exports.signUpFail = {
    type: 'object',
    properties: {
        data: {
            type: 'object',
            properties: {
                message: {},
                additionalInfo: {type: "string"}
            }
        },
        meta: {
            type: "object",
            properties: {
                code: {type: "string"},
                sourceType: {type: "string"},
                sourceName: {type: "string"},
                version: {type: "string"},
                shamsiDate: {type: "string"}
            }
        }
    }
}

exports.login = {
    type: 'object',
    properties: {
        data: {
            inArray: ['login is successful!']
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    }
}

exports.loginFail = {
    type: 'object',
    properties: {
        data: {
            type: 'object',
            properties: {
                message: {},
                additionalInfo: {type: "string"}
            }
        },
        meta: {
            type: "object",
            properties: {
                code: {type: "string"},
                sourceType: {type: "string"},
                sourceName: {type: "string"},
                version: {type: "string"},
                shamsiDate: {type: "string"}
            }
        }
    }
}

exports.logoutFail = {
    type: 'object',
    properties: {
        data: {
            type: "object",
            properties: {
                msg: {type: "string"},
            }
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    },
    required: ["data", "meta"]
}

exports.logout = {
    type: 'object',
    properties: {
        data: {
            type: "object",
            properties: {
                msg: {type: "string"},
            }
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    },
    required: ["data", "meta"]
}

exports.getProfile = {
    type: 'object',
    properties: {
        data: {
            inArray: []
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    }
}
exports.getProfileFail = {
    type: 'object',
    properties: {
        data: {
            type: 'object',
            properties: {
                message: {},
                additionalInfo: {type: "string"}
            }
        },
        meta: {
            type: "object",
            properties: {
                code: {type: "string"},
                sourceType: {type: "string"},
                sourceName: {type: "string"},
                version: {type: "string"},
                shamsiDate: {type: "string"}
            }
        }
    }
}

exports.editProfile = {
    type: 'object',
    properties: {
        data: {

        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    }
}
exports.editProfileFail = {
    type: 'object',
    properties: {
        data: {
            type: 'object',
            properties: {
                message: {},
                additionalInfo: {type: "string"}
            }
        },
        meta: {
            type: "object",
            properties: {
                code: {type: "string"},
                sourceType: {type: "string"},
                sourceName: {type: "string"},
                version: {type: "string"},
                shamsiDate: {type: "string"}
            }
        }
    }
}

exports.changePassword = {
    type: 'object',
    properties: {
        data: {
            inArray: []
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    }
}
exports.changePasswordFail = {
    type: "object",
    status: {type: 'string'},
    error: {
        type: 'object',
        properties: {
            data: {
                type: "object",
                properties: {
                    message: {type: "string"}
                }
            },
            meta: {}
        },
        required: ["data", "meta"]
    }
}

exports.getAllProfile = {
    type: "object",
    properties: {
        data: {
            type: "array",
            status: {type: 'string', default: 'ok'},
            properties: {
                result: {
                    type: 'array'
                }
            }
        },
        meta: {
            type: "object",
            properties: {
                shamsiDate: {type: "string"}
            }
        }
    },
    required: ["data", "meta"]
}
exports.getAllProfileFail = {
    type: 'object',
    status: {type: "string", default: 'error'},
    error: {
        properties: {
            data: {
                type: "object",
                properties: {
                    message: {
                        type: 'object',
                        properties: {
                            fa: {type: "string"},
                            en: {type: "string"}
                        }
                    },
                }
            },
            meta: {
                type: "object",
                properties: {
                    shamsiDate: {type: "string"}
                }
            },
        },
        required: ["data", "meta"]
    }
}
